### PR TITLE
Monad typeclass

### PR DESCRIPTION
**Pull request by felixwiemuth**
See #1648

### CODE REVIEW

**Comment by felixwiemuth** [comment]
See the discussion in #1648.

File: main/src/library/Nel.flix [hunk 1]
```diff
@@ -57,6 +57,11 @@ instance Functor[Nel] {
     pub def map(f: a -> b & ef, xs: Nel[a]): Nel[b] & ef = Nel.map(f, xs)
 }
 
+// TODO implement Applicative
+// instance Monad[Nel] {
```
Comment (magnus-madsen) [inline main/src/library/Nel.flix:61]: How are you able to implement `Monad` with an `Applicative`. Would that not be a compile-time error?

Paging @mlutze Why is this not rejected? :)

Comment (mlutze) [reply main/src/library/Nel.flix:61]: It's commented out. (Or am I missing something?)

Comment (magnus-madsen) [reply main/src/library/Nel.flix:61]: 😝

File: main/src/library/Option.flix [hunk 2]
```diff
@@ -154,6 +158,16 @@ namespace Option {
             }
         }
 
+    ///
+    /// If argument `x` is a `Some`, return the result of applying the function to the value inside `x`.
+    /// Otherwise return `None`.
+    ///
+    pub def flatMap(f: a -> Option[b] & e, x: Option[a]) : Option[b] & e =
```
Comment (magnus-madsen) [inline main/src/library/Option.flix:165]: What was wrong with the old flatMap? Which also uses the same documentations style as the rest of the library? :)

File: main/src/library/Monad.flix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/// that takes a normal value and produces a monadic value to a monadic value. That is, the bind mechanism
+/// supports extraction of monadic values, or, viewed differently, allows to combine (flatten) nested
+/// monadic values (`flapMap` can be understood as a `Functor.map` followed by a `flatten`).
+///
+class Monad[m : Type -> Type] with Applicative[m] {
+
+    ///
+    /// Apply function `f` to the monadic value `x`, resulting in a combined monadic value.
+    ///
+    pub def flatMap(f: a -> m[b] & e, x: m[a]) : m[b] & e
```
Comment (magnus-madsen) [inline main/src/library/Monad.flix:28]: Effects should be named `ef` (see STYLE.md)

**Comment by magnus-madsen** [comment]
Added minor comments. Otherwise it looks ready to merge.

Should I create a ticket to remove `map2` and `flatMap2` from everywhere in the std. lib?

File: main/src/library/Nel.flix (hunk 1 above)
Comment (magnus-madsen) [reply main/src/library/Nel.flix:61]: Welp, that was embarrassing.

**Comment by magnus-madsen** [comment]
Replying to myself: There is already a ticket, but its only for `map`.

File: main/src/library/Nel.flix (hunk 1 above)
Comment (felixwiemuth) [reply main/src/library/Nel.flix:61]: Okay, so this is resolved I guess? I just wanted to indicate that the instance implementation basically is there, but `Applicative` needs to be implemented first.

File: main/src/library/Monad.flix (hunk 3 above)
Comment (felixwiemuth) [reply main/src/library/Monad.flix:28]: Ah, good there is a style guide now. I updated it in 4 places.

**Comment by felixwiemuth** [comment]
> Should I create a ticket to remove `map2` and `flatMap2` from everywhere in the std. lib?

`flatMap2` is a different discussion, see #1648.

**Comment by felixwiemuth** [comment]
I would agree that this is ready to merge now, but you probably want to read my longer reasoning about several aspects (https://github.com/flix/flix/issues/1648#issuecomment-956172368) first.

**Comment by magnus-madsen** [comment]
Thanks!

**Comment by mlutze** [comment]
Woohoo!

**Pull request merged**